semantic_conventions/http-common.yaml

```diff
@@ -5,10 +5,51 @@ groups:
     prefix: http
     attributes:
       - id: request.method
-        type: string
+        type:
+          allow_custom_values: true
+          members:
+            - id: connect
+              value: "CONNECT"
+              brief: 'CONNECT method.'
+            - id: delete
+              value: "DELETE"
+              brief: 'DELETE method.'
+            - id: get
+              value: "GET"
+              brief: 'GET method.'
+            - id: head
+              value: "HEAD"
+              brief: 'HEAD method.'
+            - id: options
+              value: "OPTIONS"
+              brief: 'OPTIONS method.'
+            - id: patch
+              value: "PATCH"
+              brief: 'PATCH method.'
+            - id: post
+              value: "POST"
+              brief: 'POST method.'
+            - id: put
+              value: "PUT"
+              brief: 'PUT method.'
+            - id: trace
+              value: "TRACE"
+              brief: 'TRACE method.'
```

Comment:
How did we decide on this particular set of methods? Is there some specific criteria or are these just the methods the PR author thought were the most common/used? I would suggest we use the methods from the IANA HTTP spec (there are 39) at the least. The point of restricting this is to prevent infinite cardinality, but I don't see a reason to restrict to just 9.

Reply:
For the reasoning, see the `note:` below, but I think I agree with you that the largest reasonable list of methods would be best to use.

Reply:
> How did we decide on this particular set of methods?

by following what popular HTTP clients and web frameworks do in strongly typed languages (having a list of standard methods and allowing custom ones, including IANA, but not declaring them):

- Java 
  - [HttpURLConnection](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html#setRequestMethod-java.lang.String-
https://pkg.go.dev/net/http#pkg-constants) only supports predefined standard methods
  - [JDK HttpRequest.Builder](https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpRequest.Builder.html) supports even smaller subset of simple to create methods
  - [OkHttp](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-request/-builder/) - the same
  - [Netty](https://netty.io/4.1/api/io/netty/handler/codec/http/HttpMethod.html) - the same
  - [spring](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RequestMethod.html) the same
- .NET
  -  [Httpclient](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpmethod?view=net-7.0#properties) -only standard methods
  - [ASP.NET Core](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.routing.httpmethodattribute?view=aspnetcore-7.0)


none of the HTTP clients declares IANA methods. While I'm not opposed to adding them, but they all seem to be add-on rather a standard experience.

Reply:
>  The point of restricting this is to prevent infinite cardinality, but I don't see a reason to restrict to just 9.

the point is to give the instrumentation a baseline they have to support while limiting cardinality. Users of specific methods may be able to enable them one by one without allowing all non-standard IANA methods. Or we can grow this list based on demand in the spec.

39 is still pretty big and most users never need anything but 9 (relying on standard methods available in their HTTP clients and web frameworks)

Reply:
> 39 is still pretty big and most users never need anything but 9 (relying on standard methods available in their HTTP clients and web frameworks)

Whether it is big or not is totally depending on the business scenario.

Reply:
> Whether it is big or not is totally depending on the business scenario.

exactly. So support for standard methods is required, and the rest should be opt-in in one way or another.

Reply:
@lmolkova HttpUrlConnection actually seems to limit the methods. But:

Java
* JDK HttpRequest has a method() method with no such restrictions (the RFC part it references is only about the syntax of the method string, not the fixed set) <https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/HttpRequest.Builder.html#method(java.lang.String,java.net.http.HttpRequest.BodyPublisher)>
* OkHttp method() <https://square.github.io/okhttp/4.x/okhttp/okhttp3/-request/-builder/method/> -- same
* Netty: HttpMethod is not an enum, has a public constructor taking a string: <https://netty.io/4.1/api/io/netty/handler/codec/http/HttpMethod.html#HttpMethod-java.lang.String-> explicitly for custom methods
* Spring -- Don't know enough about it, but what you referenced is only an enum for use with annotations

.NET
* Httpclient: Supports nonstandard methods via public string constructor of HttpMethod (not an enum): https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpmethod.-ctor?view=net-7.0#system-net-http-httpmethod-ctor(system-string)
* ASP.NET core: Looks like you can derive new HttpMethod attributes quite trivially (base class is public) https://github.com/dotnet/aspnetcore/blob/v7.0.5/src/Mvc/Mvc.Core/src/HttpDeleteAttribute.cs

Go HTTP module: These are only string constants and in the description it already says "Common HTTP methods." (note the "Common"). You can use any custom string as well.

So in summary, only Java Spring and (legacy) HttpUrlConnection seem to actually restrict methods of those you mentioned.

Reply:
> exactly. So support for standard methods is required, and the rest should be opt-in in one way or another.

The 39 are all standard methods, they all have RFCs attached. I don't see any scenario for tracing where 39 is too large. This should easily fit in any UI listview. If 39 already counts as "high cardinality", what will we do with http.route?

Reply:
> If 39 already counts as "high cardinality", what will we do with http.route?

+1

Reply:
@Oberon00 the question is not about restriction, but about the baseline. There is no restriction in this PR, there is a set of defaults everyone has to support. The rest is custom - same way as clients I mentioned.

Reply:
Ah, that's what you meant.
But still, the conservative option here is to allow as many methods as reasonably possible out of the box, to limit breakage.

Reply:
`http.route` cardinality is controlled by the application is low-ish in practice. No one external can add new HTTP route. to server metrics except developers clearly adding new API.


We can argue about the 'standard' word. but the registry is mentioned in the [Extending HTTP/Method extensibility](https://www.rfc-editor.org/rfc/rfc9110.html#section-16.1) section. None of the IANA methods are defined as a constants in general-purpose clients, i.e. there is no HTTP stack that makes it convenient to use them. Why should we?

Reply:
To limit disruption by "other" appearing instead of the actual HTTP method.
The goal of putting in this limitation is "to prevent cardinality explosion if a malicious (or buggy) client sends custom and dynamic methods." Any method we scrub for other reasons is a bad thing, and we should limit these "incidental" conversions to "other" as much as possible.

Reply:
The last argument is that 28 (or so) of IANA methods are defined for WebDAV. If WebDAV is important to support, we'll have WebDAV extension to HTTP conventions or a separate convention. We'll see.

Here's the list of RFCs IANA refers to:
- [RFC3744](https://tools.ietf.org/html/rfc3744): The WebDAV Access Control Protocol
- [RFC3253](https://tools.ietf.org/html/rfc3253): Versioning Extensions to WebDAV
- [RFC5842](https://tools.ietf.org/html/rfc5842): Binding Extensions to Web Distributed Authoring and Versioning (WebDAV)
- [RFC9110](https://tools.ietf.org/html/rfc9110): Binding HTTP Requests to Application-Provided Secure Elements
- [RFC4918](https://tools.ietf.org/html/rfc4918): HTTP Extensions for Web Distributed Authoring and Versioning (WebDAV)
- [RFC2068](https://tools.ietf.org/html/rfc2068): Hypertext Transfer Protocol -- HTTP/1.1
- [RFC4791](https://tools.ietf.org/html/rfc4791): Calendaring Extensions to WebDAV (CalDAV)
- [RFC8144](https://tools.ietf.org/html/rfc8144): Use of the Prefer Header Field in Web Distributed Authoring and Versioning (WebDAV)
- [RFC5689](https://tools.ietf.org/html/rfc5689): Extended MKCOL for Web Distributed Authoring and Versioning (WebDAV)
- [RFC4437](https://tools.ietf.org/html/rfc4437): Web Distributed Authoring and Versioning (WebDAV) Redirect Reference Resources
- [RFC3648](https://tools.ietf.org/html/rfc3648): Web Distributed Authoring and Versioning (WebDAV) Ordered Collections Protocol
- [RFC5789](https://tools.ietf.org/html/rfc5789): PATCH Method for HTTP
- [RFC9113](https://tools.ietf.org/html/rfc9113): Indicating the Media Feature Requirements of a Printer Using IPP Attributes
- [RFC5323](https://tools.ietf.org/html/rfc5323): Web Distributed Authoring and Versioning (WebDAV) SEARCH


As you can see, IANA is HTTP and WebDAV, but there are other popular methods like PURGE that are not there. I.e. 39  won't be enough either. 

> To limit disruption by "other" appearing instead of the actual HTTP method.

the disruption of `other` is quite limited:
1. you have a non-standard method
2. you add otel, you see `other` methods
3.  you opt-in your explicit set of non-common methods, not including all 39.

otherwise, you risk having ~30x time series in your metrics that contain long tail of buggy clients and attacks

Reply:
> otherwise, everyone risks having ~30x time series in their metrics that contain long tail of buggy clients and attacks

Not sure what you mean by "otherwise", the suggestion of 39 vs 9 is a bit more than 4x.

Reply:
@Oberon00 any thoughts on WebDAV? Do you feel it's essential to provide first-class support by default? 

> Not sure what you mean by "otherwise", the suggestion of 39 vs 9 is a bit more than 4x.
I meant that if we allow 39 methods, 30 of them won't be used by most systems, but will be open to potentially attackers.

Adding 30 more possible attribute values can totally lead to hitting cardinality limit and causing 'good' time series to be dropped.

Reply:
> Adding 30 more possible attribute values can totally lead to hitting cardinality limit and causing 'good' time series to be dropped.

Relative to the pre-PR situation, we are not adding more with these 39, we are reducing less.

Reply:
is reducing less to make WebDAV work worth it? maybe the fact that http://www.webdav.org/ recent news is from 2009 (or the `http://` in front of it can change your opinion?)

Reply:
In the end, this point is not very important for me, we can agree to disagree. It would be good if we could clarify that adding more methods to these defaults in spec later is not a breaking change.

Reply:
cool! enum values are explicitly called out as something that can change.

https://github.com/open-telemetry/opentelemetry-specification/blame/37ab6b056172ad36d31ce217927e47bbe031d831/specification/versioning-and-stability.md#L221-L224

```text
Things not listed in the above are not expected to remain stable via semantic
convention and are allowed (or expected) to change. A few examples:

- The values of attributes
  - Specifically for `enums` the list of allowed values is expected to change
    overtime.
  - Even for `enums` that limit allowed values to semconv, some may need to
    updated values in the future. Tooling should expect unknown values.
```

So I'll resolve this discussion with the agreement that we can always add other popular methods as we get more feedback from instrumentation authors and users.